synthesizer/process/src/stack/finalize_types/initialize.rs

```diff
@@ -329,16 +355,45 @@ impl<N: Network> FinalizeTypes<N> {
     fn check_get_or_use(
         &mut self,
         stack: &(impl StackMatches<N> + StackProgram<N>),
-        finalize_name: &Identifier<N>,
         get_or_use: &GetOrUse<N>,
     ) -> Result<()> {
-        // Ensure the declared mapping in `get.or_use` is defined in the program.
-        if !stack.program().contains_mapping(get_or_use.mapping_name()) {
-            bail!("Mapping '{}' in '{}/{finalize_name}' is not defined.", get_or_use.mapping_name(), stack.program_id())
-        }
-        // Retrieve the mapping from the program.
-        // Note that the unwrap is safe, as we have already checked the mapping exists.
-        let mapping = stack.program().get_mapping(get_or_use.mapping_name()).unwrap();
+        // Retrieve the mapping.
+        let mapping = match get_or_use.mapping() {
+            MappingLocator::Locator(locator) => {
+                // Retrieve the program ID.
+                let program_id = locator.program_id();
+                // Retrieve the mapping_name.
+                let mapping_name = locator.resource();
+
+                // Ensure the locator does not reference the current program.
+                if stack.program_id() == program_id {
+                    bail!("Locator '{locator}' does not reference an external mapping.");
+                }
+                // Ensure the current program contains an import for this external program.
+                if !stack.program().imports().keys().contains(program_id) {
+                    bail!("External program '{locator}' is not imported by '{program_id}'.");
+                }
+                // Retrieve the program.
+                let external = stack.get_external_program(program_id)?;
+                // Ensure the mapping exists in the program.
+                if !external.contains_mapping(mapping_name) {
+                    bail!("Mapping '{mapping_name}' in '{program_id}' is not defined.")
+                }
+                // Retrieve the mapping from the program.
+                // Note that the unwrap is safe, as we have already checked the mapping exists.
+                external.get_mapping(mapping_name).unwrap()
```

Comment:
Even if it is safe, what benefit is the unwrap here?

Reply:
Good catch, they are unnecessary. Replacing with `?`s.